[DXIL] `exp`, `any`, `lerp`, & `rcp` Intrinsic Lowering

farzonl opened the pull request:
This change implements lowering for #70076, #70100, #70072, & #70102 
`CGBuiltin.cpp` - - simplify `lerp` intrinsic
`IntrinsicsDirectX.td` - simplify `lerp` intrinsic
`SemaChecking.cpp` - remove unnecessary check
`DXILIntrinsicExpansion.*` - add intrinsic to instruction expansion cases 
`DXILOpLowering.cpp` -  make sure `DXILIntrinsicExpansion` happens first 
`DirectX.h` - changes to support new pass
`DirectXTargetMachine.cpp` - changes to support new pass

Why `any`, and `lerp` as instruction expansion just for DXIL?
-  SPIR-V there is an [OpAny](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpAny)
- SPIR-V has a GLSL lerp extension via [Fmix](https://registry.khronos.org/SPIR-V/specs/1.0/GLSL.std.450.html#FMix)

Why `exp` instruction expansion?
- We have an `exp2` opcode and `exp` reuses that opcode. So instruction expansion is a convenient way to do preprocessing. 
- Further SPIR-V has a GLSL exp extension via  [Exp](https://registry.khronos.org/SPIR-V/specs/1.0/GLSL.std.450.html#Exp) and [Exp2](https://registry.khronos.org/SPIR-V/specs/1.0/GLSL.std.450.html#Exp2) 

Why `rcp` as instruction expansion?
This one is a bit of the odd man out and might have to move to `cgbuiltins` when we better understand SPIRV requirements. However I included it because it seems like [fast math mode has an AllowRecip flag](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_fp_fast_math_mode) which lets you compute the reciprocal without performing the division.  We don't have that in DXIL so thought to include it.

bogner commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -0,0 +1,187 @@
+#include "llvm/Support/ErrorHandling.h"
+
+#define DEBUG_TYPE "dxil-intrinsic-expansion"
+#define M_LOG2E_F 1.44269504088896340735992468100189214f
```
You can get this from `llvm/Support/MathExtras.h` as `numbers::log2e`

bogner commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -0,0 +1,187 @@
+      Ty->isVectorTy()
+          ? ConstantVector::getSplat(
+                ElementCount::getFixed(
+                    dyn_cast<FixedVectorType>(Ty)->getNumElements()),
```
If we aren't checking for null this should be `cast` not `dyn_cast`. Does this do something bad if we somehow get here with scalable vectors?

farzonl replied:
I'l change the cast and investigate other vector types. I haven't tested with scalable vectors. I didn't see a generic vector type I could use and I just noticed in lldb the clang vectors had a FixedVector type so went with that.

bogner commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -0,0 +1,187 @@
+  return true;
+}
+
+static bool expandReciprocalIntrinsic(CallInst *Orig) {
```
For consistency this should probably just be expandRcpIntrinsic - we didn't spell "Exp" as "Exponential" in the other function above

bogner commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -0,0 +1,187 @@
+  return false;
+}
+
+static bool intrinsicExpansion(Module &M) {
```
Best practice for function names is to make them "verb-y". I'd call this "expandIntrinsics".

bogner commented on llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp:
```diff
@@ -0,0 +1,187 @@
+static bool intrinsicExpansion(Module &M) {
+  for (auto &F : make_early_inc_range(M.functions())) {
+    if (!isIntrinsicExpansion(F))
+      continue;
```
`isIntrinsicExpansion` needs to duplicate the list of opcodes we care about - are we really gaining anything over simply calling `expandIntrinsic` and letting it fall through to the "do nothing" case?

farzonl replied:
my thinking was if we new earlier we could avoid a uses loop.

bogner commented on llvm/test/CodeGen/DirectX/any.ll:
```diff
@@ -0,0 +1,133 @@
+
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
+target triple = "dxil-pc-shadermodel6.7-library"
```
You can probably strip the datalayout from these tests, and maybe even the triple

bogner commented on llvm/test/CodeGen/DirectX/any.ll:
```diff
@@ -0,0 +1,133 @@
+
+
+; CHECK:icmp ne i1 %{{.*}}, false
+; Function Attrs: noinline nounwind optnone
```
Usually best to strip the function attr comments from tests

bogner commented on llvm/test/CodeGen/DirectX/any.ll:
```diff
@@ -0,0 +1,133 @@
+
+; CHECK:icmp ne i1 %{{.*}}, false
+; Function Attrs: noinline nounwind optnone
+define noundef i1 @any_bool(i1 noundef %p0) #0 {
```
It looks like the `attributes #0 = ...` definition is stripped, but you've left the `#0` on these functions. If we don't need the attributes for the test these can just be stripped.

bogner commented on llvm/test/CodeGen/DirectX/any.ll:
```diff
@@ -0,0 +1,133 @@
+}
+
+; Function Attrs: nocallback nofree nosync nounwind willreturn
+declare i1 @llvm.dx.any.i1(i1) #1
```
It doesn't really matter, but I find it a bit tidier to put all of these function declarations together at the end.

farzonl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -5254,7 +5250,11 @@ bool CheckAllArgsHaveFloatRepresentation(Sema *S, CallExpr *TheCall) {
+                       ? S->Context.HalfTy
+                       : S->Context.FloatTy;
+    if (PassedType == S->Context.DoubleTy ||
+        !PassedType->hasFloatingRepresentation()) {
```
1. This was originally just a check to make sure args were floats and I shoehorned in the double case to handle the type promotions caused by `DefaultVariadicArgumentPromotion` when calling the builtin directly. There are no type promotions for BFloat16 or Ibm128  and we don't define those types in hlsl so i'm not able to test for these other floating point types.
2. The name should probably change.
3. I was hoping to avoid triggering the error on api calls of half\vector<half>  ie. lerp(half,half,half).  There is likely a better way for me to structure this to make it look more correct.
4. It can't be tested I added a test for it on line 91 of `lerp-builtin.hlsl` and the original type is lost by the time we call the builtin. I added a note to explain why it was `float` and not  `half`.

farzonl replied:
I have a solution I think will make things more clear.

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18015,38 +18015,11 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
-           XVecTy->getElementType() == YVecTy->getElementType() &&
-           XVecTy->getElementType() == SVecTy->getElementType() &&
-           "Lerp requires float vectors to be of the same type.");
+    if (!E->getArg(0)->getType()->hasFloatingRepresentation())
```
`X->getType()` is a `Type` pointer. `E->getArg(0)->getType()` is a `QualType`. Only QualTypes allow us to use `hasFloatingRepresentation`

bogner commented on llvm/test/CodeGen/DirectX/any.ll:
```diff
@@ -0,0 +1,105 @@
+; Make sure dxil operation function calls for any are generated for float and half.
+
+; CHECK:icmp ne i1 %{{.*}}, false
+define noundef i1 @any_bool(i1 noundef %p0) {
```
Probably worth adding a `; CHECK-LABEL: define {{.*}} @any_bool` or so to make sure these checks are looking at the right functions and also potentially improve the error message if they fail

bogner commented on llvm/test/CodeGen/DirectX/any.ll:
```diff
@@ -0,0 +1,105 @@
+
+; Make sure dxil operation function calls for any are generated for float and half.
+
+; CHECK:icmp ne i1 %{{.*}}, false
```
Folks usually put a space between `CHECK:` and the thing being checked (`CHECK: icmp ...`). I find that slightly more readable.

farzonl replied:
That was my thinking as well, but all the other tests in this directory lacked the space and I didn't want to go against the the grain.

bogner reviewed (approved):
A couple more very minor notes on the tests, then this LGTM

farzonl committed:
[DXIL] exp, any, lerp, & rcp Intrinsic Lowering
This change implements lowering for #70076, #70100, #70072, & #70102
`CGBuiltin.cpp` - - simplify `lerp` intrinsic
`IntrinsicsDirectX.td` - simplify `lerp` intrinsic
`SemaChecking.cpp` - remove unnecessary check
`DXILIntrinsicExpansion.*` - add intrinsic to instruction expansion cases
`DXILOpLowering.cpp` -  make sure `DXILIntrinsicExpansion` happens first
`DirectX.h` - changes to support new pass
`DirectXTargetMachine.cpp` - changes to support new pass

Why `any`, and `lerp` as instruction expansion just for DXIL?
-  SPIR-V there is an [OpAny](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpAny)
- SPIR-V has a GLSL lerp extension via [Fmix](https://registry.khronos.org/SPIR-V/specs/1.0/GLSL.std.450.html#FMix)

Why `exp` instruction expansion?
- We have an `exp2` opcode and `exp` reuses that opcode. So instruction expansion is a convenient way to do preprocessing.
- Further SPIR-V has a GLSL exp extension via  [Exp](https://registry.khronos.org/SPIR-V/specs/1.0/GLSL.std.450.html#Exp) and [Exp2](https://registry.khronos.org/SPIR-V/specs/1.0/GLSL.std.450.html#Exp2)

Why `rcp` as instruction expansion?
This one is a bit of the odd man out and might have to move to `cgbuiltins` when we better understand SPIRV requirements. However I included it because it seems like [fast math mode has an AllowRecip flag](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_fp_fast_math_mode) which lets you compute the reciprocal without performing the division.  We don't have that in DXIL so thought to include it.

farzonl committed:
add labels for test cases

farzonl committed:
enforce float\half type checks

coopp reviewed (approved):
Looks good to me for what I can understand.
